be/src/exec/vectorized/aggregate/agg_hash_map.h

```diff
@@ -80,10 +80,27 @@ struct AggHashMapWithOneNumberKey {
         DCHECK(!key_columns[0]->is_nullable());
         auto column = down_cast<ColumnType*>(key_columns[0].get());
 
+        AggDataPtr* agg_buf = agg_states->data();
+
+        // precompute hash value first, save it into `agg_states` array.
+        static_assert(sizeof(AggDataPtr) == sizeof(size_t));
+        size_t* hash_values = reinterpret_cast<size_t*>(agg_buf);
         for (size_t i = 0; i < column->size(); i++) {
             FieldType key = column->get_data()[i];
-            auto iter = hash_map.lazy_emplace(key, [&](const auto& ctor) { ctor(key, allocate_func()); });
-            (*agg_states)[i] = iter->second;
+            size_t hashval = hash_map.hash_function()(key);
+            hash_values[i] = hashval;
+        }
+
+        static const size_t prefetch = 16;
+        // during insert, we do prefetch and insert with hash.
+        for (size_t i = 0, j = prefetch; i < column->size(); i++, j++) {
+            if (j < column->size()) {
```

Comment:
I think we could remove this if condition by split the loop

Reply:
OK, but I suspect that will make code complicated(since we are going to apply this technique to other agg hash map instances) . And I have checked profile data, `branch-misses` is pretty small enough to be neglected.

Reply:
to look closer to `branch-misses` :
1. `build hash map` takes 6.64% 
2. and `j<column->size()` takes  1.43%

which means `branch-misses` of that code is pretty low enough. We don't need to strive to save a little branch miss to make code complicated. 

![image](https://user-images.githubusercontent.com/1081215/135173778-750c1f78-1de0-4e4b-85b6-6d5bfde8d7d8.png)

![image](https://user-images.githubusercontent.com/1081215/135174048-f2c10f2d-7c64-4153-945e-97f3dfac2c61.png)

Reply:
Looks like there is register pressure. I guess

- `-0x50(%rbp)` is `__prefetch_index`
- `-0x60(%rbp)` is `column_size`

Ideally `column_size` should be put into the register instead of the stack.  Probably we are doing a lot of inline functions that register pressure is inevitable.